docs/nodemcu-lrm.md

```diff
@@ -0,0 +1,331 @@
+
+The NodeMCU runtime implements a non-blocking threaded model that is similar to that of `node.js`, and hence most Lua execution is initiated from C event-triggered callback (CB) routines.  NodeMCU allow full error diagnostics to be recovered from CB unprotected errors by including an additional auxiliary library function that prints a full traceback before triggering processor restart.
+
+### Additional API Functions and Types
+
+These are available for developers coding new library modules in C.  Note that if you compare the standard and NodeMCU versions of `lua.h` you will find a small number of entries not listed below.  This is because the Lua 5.1 and Lua 5.3 variants are incompatible owing to architectural differences.  However, `laubxlib.h` includes equivalent wrapper version-compatible functions that may be used safely for both versions.
+
+#### ROTable and ROTable_entry
+
+Extra structure types used in `LROT` macros to declare static RO tables.  See detailed section below.
+
+#### lua_createrotable
+
+` void (lua_createrotable) (lua_State *L, ROTable *t, const ROTable_entry *e, ROTable *mt);`         [-0, +1, -]
+
+Create a RAM based `ROTable` pointing to the `ROTable_entry` vector `e`, and metatable `mt`.
+
+#### lua_debugbreak and ASSERT
+
+`  void (lua_debugbreak)(void);`
+
+`lua_debugbreak()` and `ASSERT(condtion)` are available for for development debugging.  If `DEVELOPMENT_USE_GDB` is defined then these will trigger a debugger break and evaluate a conditional assert prologue on the same.  If not, then these are effectively ignored and generate no executable code.
+
+#### lua_dump
+
+` int lua_dump (lua_State *L, lua_Writer writer, void *data, int strip);`         [-0, +0, –]
+
+Dumps function at the top of the stack function as a binary chunk as per LRM.  However the last argument is now an integer is in the range -1..2, rather a boolean as per standard Lua:
+- `strip`
+    - -1, use the current default strip level (which can be set by [`lua_striptdebug`](#lua_striptdebug))
+    - 0, keep all debug info
+    - 1, discard Local and Upvalue debug info; keep line number info
+    - 2, discard Local, Upvalue and line number debug info
+
+The internal NodeMCU `Proto` encoding of debug line number information is typically 15× more compact than in standard Lua; the intermediate `strip=1` argument allows the removal of must of the debug information whilst retaining the ability to produce a proper line number traceback on error. 
+
+#### lua_freeheap
+
+` int (lua_freeheap) (void);`         [-0, +0, –]
+
+returns the amount of free heap available to the Lua memory allocator.
```

Comment:
Maybe make a note that this doesn't take into account fragmentation of said available memory? I.e. the max allocatable chunk may be significantly smaller.

Reply:
Agreed, but its the best we have ATM.  This is also a very cheap call since this value is held as a `size_t` in DRAM0, so this is just a getter for this.